### ISSUE TITLE
Sequentially training two models end to end

### ISSUE THREAD

**Issue by varunjain3** [OpenNMT/OpenNMT-py#1922]
Hey, We are doing a project based on transformers and using open-nmt to train our transformer models.
Is there a way one can train 2 models sequentially some sort of GAN based approach?

**Comment by francoishernandez** [comment]
Hi,

Such case is not handled in the existing codebase. It should not be very hard to adapt though. For instance you could build two models in `train_single`, and adapt the `trainer` and a few other parts accordingly.

**Comment by varunjain3** [comment]
@francoishernandez Thanks for the heads up.
I m new to these kinds of big APIs. I understand that I'll need to initialise both models by adapting the train_single. But I m not sure how should I give first model's output as the input to the second model. What all should I change?

**Comment by francoishernandez** [comment]
Model forward is performed here:
https://github.com/OpenNMT/OpenNMT-py/blob/dfdd4554cf1c9327c2c249ab11a978efe72a93d9/onmt/trainer.py#L366

You can adapt this to do it on both your models sequentially.

As for freezing one of them, you can probably switch it to `eval` mode by calling `model.eval()`.

**Comment by varunjain3** [comment]
Hi @francoishernandez 
I checked this, but I think the F-prop you have pointed is the validation check, I want to add loss for each train iteration, is there a way I can do that? or do I have to call this second model for validations steps only?

**Comment by francoishernandez** [comment]
No, the validation forward is here:
https://github.com/OpenNMT/OpenNMT-py/blob/dfdd4554cf1c9327c2c249ab11a978efe72a93d9/onmt/trainer.py#L318

**Comment by varunjain3** [comment]
So the src and tgt at this stage are already converted into embeddings
https://github.com/OpenNMT/OpenNMT-py/blob/dfdd4554cf1c9327c2c249ab11a978efe72a93d9/onmt/trainer.py#L366-L368

Is there a way I could remap them back without having too much time penalty?

**Comment by francoishernandez** [comment]
Not sure why you would want that at this point.

**Comment by varunjain3** [comment]
As discussed above, ill need them for feed forward from my second model. Is there a way to get them?

**Comment by francoishernandez** [comment]
> Is there a way one can train 2 models sequentially some sort of GAN based approach?

If you want to train those sequentially, your second model's input is your first model output. 

Anyways, `batch.src` and `batch.tgt` are IDs here if you need them. Embeddings are performed as first step of the encoder.

**Comment by varunjain3** [comment]
@francoishernandez Thanks for the idea!
So as you said, I found a way to adapt my output into natural language.

So now I have my Input for the second model.
But I m not sure how should I pass it through the second model.

https://github.com/OpenNMT/OpenNMT-py/blob/bc95e03875aabf4363161aac70830ecbdb762d91/onmt/bin/translate.py#L18-L32

I tried making a translator object. But I was stuck as I realised that the src_shard is not a sentence but a generator class. Would you suggest me to make a trainer object and put it in eval mode? Is there an easier way or some API I could use to enter input and get predictions?
